src/CacheManager.Core/Internal/CacheKeysHelper.cs

```diff
@@ -0,0 +1,115 @@
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace CacheManager.Core.Internal
+{
+    /// <summary>
+    /// Helper functions for manipulating keys.
+    /// </summary>
+    static public class CacheKeysHelper
+    {
+        /// <summary>
+        /// Filter a list of keys based on pattern. For handles that already have a list of keys and don't have service side filtering.
+        /// </summary>
+        /// <param name="keys"></param>
+        /// <param name="pattern">Glob to filter the key strings with</param>
+        /// <returns></returns>
+        public static IEnumerable<string> FilterBy(this IEnumerable<string> keys, string pattern)
+        {
+            switch (MatchType(pattern))
+            {
+                case KeyMatchPattern.ALL:
+                    return keys;
+                case KeyMatchPattern.CONTAINS_STRING:
+                    return keys.Where(k => k.Contains(pattern));
+                case KeyMatchPattern.ENDS_WITH_CHAR:
+                    return keys.Where(k => k.Length == pattern.Length + 1 && k.StartsWith(pattern));
+                case KeyMatchPattern.ENDS_WITH_WILDCARD:
+                    return keys.Where(k => k.StartsWith(pattern));
+                case KeyMatchPattern.STARTS_WITH_CHAR:
+                    return keys.Where(k => k.Length == pattern.Length + 1 && k.EndsWith(pattern));
+                case KeyMatchPattern.STARTS_WITH_WILDCARD:
+                    return keys.Where(k => k.EndsWith(pattern));
+                case KeyMatchPattern.CONTAINS_SINGLE_CHAR:
+                    {
+                        var patterns = pattern.Split('?');
+                        return keys.Where(k => k.Length == pattern.Length + 1 && k.StartsWith(patterns[0]) && k.EndsWith(patterns[1]));
+                    }
+                case KeyMatchPattern.CONTAINS_SINGLE_WILDCARD:
+                    {
+                        var patterns = pattern.Split('?');
+                        return keys.Where(k => k.StartsWith(patterns[0]) && k.EndsWith(patterns[1]));
+                    }
+                default:
+                    {
+                        var regexPattern = "^" + Regex.Escape(pattern).Replace(@"\*", ".*").Replace(@"\?", ".") + "$";
+                        var regex = new Regex(regexPattern);
+                        return keys.Where(k => regex.IsMatch(k));
+                    }
+            }
+        }
+
+        static KeyMatchPattern MatchType(string pattern)
+        {
+            if (pattern == "*")
+                return KeyMatchPattern.ALL;
+
+            var firstCharMatch = pattern.IndexOf('?');
+
+            if (firstCharMatch == -1)
+            {
+                var firstWildcard = pattern.IndexOf('*');
+                if (firstWildcard == -1)
+                    return KeyMatchPattern.CONTAINS_STRING;
+
+                var lastWildcard = pattern.LastIndexOf('*');
+                if (firstWildcard == lastWildcard)
```

Comment:
It would only be common in region-less cases on the in-memory cache. Though checking the region prefix would be another option for optimization if someone hits a use case for it. It's fairly easy to extend.